### PR TITLE
Add `splitTestsCompilation` solidity setting (3): Compilation cache updates

### PR DESCRIPTION
**Pull request by alcuadrado**
This PR updates the compilation cache so that it works with the new setting.

Test failures are expected, as `build` and `test solidity` haven't been updated yet.

### CODE REVIEW

File: packages/hardhat/test/internal/builtin-plugins/solidity/build-system/partial-compilation/output-layout-cache.ts [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+      // case produces a cache hit. The invalidation matters for test roots
+      // (tested separately below).
+      const hreUnified = await getHREWithSplitConfig(project.path, false);
+      const unifiedResult = await hreUnified.solidity.build([filePath], {
+        quiet: true,
+      });
+      assert(
+        hreUnified.solidity.isSuccessfulBuildResult(unifiedResult),
+        "Unified mode build should succeed",
+      );
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat/test/internal/builtin-plugins/solidity/build-system/partial-compilation/output-layout-cache.ts:232]: good point, addressed it

File: packages/hardhat/test/internal/builtin-plugins/solidity/build-system/partial-compilation/output-layout-cache.ts [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+      assert(
+        hre.solidity.isSuccessfulBuildResult(result),
+        "Build should succeed",
+      );
+      assert.equal(
+        result.get(filePath)?.type,
+        FileBuildResultType.BUILD_SUCCESS,
+        "Should recompile when output layout fields are missing from cache",
+      );
+    });
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat/test/internal/builtin-plugins/solidity/build-system/partial-compilation/output-layout-cache.ts:395]: good point

**Review by schaable — APPROVED** [review]

File: packages/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts [hunk 3]
```diff
@@ -641,6 +642,26 @@ export class SolidityBuildSystemImplementation implements SolidityBuildSystem {
...
+        rootFile,
+        options?.scope ?? "contracts",
+      );
+
+      if (
+        cacheResult.artifactsDirectory === undefined ||
+        cacheResult.emitsTypeDeclarations === undefined ||
+        cacheResult.artifactsDirectory !== expectedLayout.artifactsDirectory ||
+        cacheResult.emitsTypeDeclarations !==
+          expectedLayout.emitsTypeDeclarations
```
Comment (schaable) [inline packages/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts:659]: Can you help me better understand why `emitsTypeDeclarations` is necessary in the cache output layout? I get that toggling unified compilation changes the `artifactsDirectory` for test roots, but isn’t `emitsTypeDeclarations` always true for contract roots and false for test roots? Is this mainly to make the output description more complete and defensively cover edge cases?

Comment (alcuadrado) [reply packages/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts:659]: I think we could re-filter by scope here instad of save it, yes. But I feel this makes it more self contained.
